File: packages/trust-receipts/src/trust-receipt.ts
/**
 * TrustReceipt - Cryptographically signed AI interaction records
 *
 * Each receipt captures:
 * - The prompt/input sent to the AI (hashed)
 * - The response/output received (hashed)
 * - User-defined attestation scores
 * - Hash chain to previous receipt
 * - Ed25519 digital signature
 *
 * Uses JSON Canonicalization Scheme (RFC 8785) for deterministic hashing.
 */

import { canonicalize } from 'json-canonicalize';

import { sha256, sign, verify, bytesToHex, hexToBytes } from './crypto';

/**
 * Attestation scores for the AI interaction.
 * All scores are user-defined floats between 0 and 1.
 * Common scores: clarity, integrity, quality, safety, accuracy.
 */
export interface Scores {
  [key: string]: number;
}

/**
 * Data required to create a TrustReceipt
 */
export interface TrustReceiptData {
  /** Unique session identifier */
  sessionId: string;
  /** The prompt/input sent to the AI (will be hashed) */
  prompt: unknown;
  /** The response/output from the AI (will be hashed) */
  response: unknown;
  /** Attestation scores (user-defined, 0-1 range) */
  scores: Scores;
  /** Optional agent/model identifier */
  agentId?: string;
  /** Optional hash of previous receipt for chaining */
  prevReceiptHash?: string;
  /** Optional custom metadata */
  metadata?: Record<string, unknown>;
  /** Include full prompt/response content in receipt (default: false, hashes only) */
  includeContent?: boolean;
}

/**
 * A signed TrustReceipt ready for storage or transmission
 */
export interface SignedReceipt {
  /** Schema version */
  version: string;
  /** ISO 8601 timestamp (UTC) */
  timestamp: string;
  /** Session identifier */
  sessionId: string;
  /** Agent/model identifier */
  agentId: string | null;
  /** SHA-256 hash of the prompt/input */
  promptHash: string;
  /** SHA-256 hash of the response/output */
  responseHash: string;
  /** Attestation scores */
  scores: Scores;
  /** Hash of previous receipt (if chained) */
  prevReceiptHash: string | null;
  /** SHA-256 hash of this receipt's canonical payload */
  receiptHash: string;
  /** Ed25519 signature (hex) */
  signature: string;
  /** Custom metadata */
  metadata: Record<string, unknown>;
  /** Full prompt content (only present if includeContent: true) */
  promptContent?: unknown;
  /** Full response content (only present if includeContent: true) */
  responseContent?: unknown;
}

/**
 * TrustReceipt class for creating and verifying signed receipts
 */
export class TrustReceipt {
  readonly version = '1.0';
  readonly timestamp: string;
  readonly sessionId: string;
  readonly agentId: string | null;
  readonly promptHash: string;
  readonly responseHash: string;
  readonly scores: Scores;
  readonly prevReceiptHash: string | null;
  readonly metadata: Record<string, unknown>;
  readonly receiptHash: string;
  readonly promptContent?: unknown;
  readonly responseContent?: unknown;

  private _signature: string = '';

  constructor(data: TrustReceiptData) {
    this.timestamp = new Date().toISOString();
    this.sessionId = data.sessionId;
    this.agentId = data.agentId ?? null;
    this.promptHash = this.hashContent(data.prompt);
    this.responseHash = this.hashContent(data.response);
    this.scores = { ...data.scores };
    this.prevReceiptHash = data.prevReceiptHash ?? null;
    this.metadata = data.metadata ?? {};

    if (data.includeContent) {
      this.promptContent = data.prompt;
      this.responseContent = data.response;
    }

    this.receiptHash = this.computeReceiptHash();
  }

  /**
   * Hash any content using SHA-256 with JSON canonicalization
   */
  private hashContent(content: unknown): string {
    if (content === undefined || content === null) {
      return sha256('');
    }
    const canonical = canonicalize(content);
    return sha256(canonical);
  }

  /**
   * Compute SHA-256 hash of the full receipt payload
   * Uses JSON Canonicalization Scheme (RFC 8785) for determinism
   */
  private computeReceiptHash(): string {
    const payload = {
      version: this.version,
      timestamp: this.timestamp,
      sessionId: this.sessionId,
      agentId: this.agentId,
      promptHash: this.promptHash,
      responseHash: this.responseHash,
      scores: this.scores,
      prevReceiptHash: this.prevReceiptHash,
      metadata: this.metadata,
    };

    const canonical = canonicalize(payload);
    return sha256(canonical);
  }

  /**
   * Sign the receipt with an Ed25519 private key
   *
   * @param privateKey - 32-byte Ed25519 private key
   */
  async sign(privateKey: Uint8Array): Promise<void> {
    const message = hexToBytes(this.receiptHash);
    const signature = await sign(message, privateKey);
    this._signature = bytesToHex(signature);
  }

  /**
   * Verify the receipt's signature
   *
   * @param publicKey - 32-byte Ed25519 public key
   * @returns true if signature is valid
   */
  async verify(publicKey: Uint8Array): Promise<boolean> {
    if (!this._signature) {
      return false;
    }

    try {
      const message = hexToBytes(this.receiptHash);
      const signature = hexToBytes(this._signature);
      return await verify(signature, message, publicKey);
    } catch {
      return false;
    }
  }

  /**
   * Check if this receipt chains correctly from a previous receipt
   *
   * @param previous - The previous receipt in the chain
   * @returns true if chain is valid
   */
  verifyChain(previous: TrustReceipt | SignedReceipt): boolean {
    const prevHash = (previous as SignedReceipt).receiptHash;
    return this.prevReceiptHash === prevHash;
  }

  /**
   * Get the signature (hex string)
   */
  get signature(): string {
    return this._signature;
  }

  /**
   * Check if receipt is signed
   */
  get isSigned(): boolean {
    return this._signature.length > 0;
  }

  /**
   * Export as JSON object
   */
  toJSON(): SignedReceipt {
    const json: SignedReceipt = {
      version: this.version,
      timestamp: this.timestamp,
      sessionId: this.sessionId,
      agentId: this.agentId,
      promptHash: this.promptHash,
      responseHash: this.responseHash,
      scores: this.scores,
      prevReceiptHash: this.prevReceiptHash,
      receiptHash: this.receiptHash,
      signature: this._signature,
      metadata: this.metadata,
    };

    if (this.promptContent !== undefined) {
      json.promptContent = this.promptContent;
    }
    if (this.responseContent !== undefined) {
      json.responseContent = this.responseContent;
    }

    return json;
  }

  /**
   * Create a TrustReceipt from JSON data
   *
   * @param data - Previously serialized receipt
   */
  static fromJSON(data: SignedReceipt): TrustReceipt {
    // Create with dummy data, then override
    const receipt = Object.create(TrustReceipt.prototype) as TrustReceipt;

    // Assign readonly properties
    (receipt as any).version = data.version;
    (receipt as any).timestamp = data.timestamp;
    (receipt as any).sessionId = data.sessionId;
    (receipt as any).agentId = data.agentId;
    (receipt as any).promptHash = data.promptHash;
    (receipt as any).responseHash = data.responseHash;
    (receipt as any).scores = data.scores;
    (receipt as any).prevReceiptHash = data.prevReceiptHash;
    (receipt as any).metadata = data.metadata;
    (receipt as any).receiptHash = data.receiptHash;
    if (data.promptContent !== undefined) {
      (receipt as any).promptContent = data.promptContent;
    }
    if (data.responseContent !== undefined) {
      (receipt as any).responseContent = data.responseContent;
    }
    receipt._signature = data.signature;

    return receipt;
  }
}

// Re-export old types for backwards compatibility during transition
/** @deprecated Use Scores instead */
export type QualityMetrics = Scores;

File: packages/trust-receipts-python/sonate/trust_receipt.py
"""
TrustReceipt - Cryptographically signed AI interaction records
"""

from dataclasses import dataclass, asdict, field
from datetime import datetime, timezone
from typing import Any, Dict, Optional, Union
from .crypto import canonicalize_json, canonicalize_and_hash, sign, verify, hex_to_bytes

Scores = Dict[str, float]


@dataclass
class TrustReceiptData:
    """Data required to create a TrustReceipt"""
    session_id: str
    prompt: Any
    response: Any
    scores: Scores
    agent_id: Optional[str] = None
    prev_receipt_hash: Optional[str] = None
    metadata: Dict[str, Any] = field(default_factory=dict)
    include_content: bool = False


@dataclass
class SignedReceipt:
    """A signed TrustReceipt ready for storage or transmission"""
    version: str
    timestamp: str
    session_id: str
    agent_id: Optional[str]
    prompt_hash: str
    response_hash: str
    scores: Scores
    prev_receipt_hash: Optional[str]
    receipt_hash: str
    signature: str
    metadata: Dict[str, Any]
    prompt_content: Optional[Any] = None
    response_content: Optional[Any] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        d = asdict(self)
        if self.prompt_content is None:
            del d['prompt_content']
        if self.response_content is None:
            del d['response_content']
        return d


class TrustReceipt:
    """TrustReceipt class for creating and verifying signed receipts"""
    
    VERSION = "1.0"
    
    def __init__(self, data: TrustReceiptData):
        """Create a new TrustReceipt"""
        self.version = self.VERSION
        self.timestamp = datetime.now(timezone.utc).isoformat()
        self.session_id = data.session_id
        self.agent_id = data.agent_id or None
        self.prompt_hash = self._hash_content(data.prompt)
        self.response_hash = self._hash_content(data.response)
        self.scores = dict(data.scores)
        self.prev_receipt_hash = data.prev_receipt_hash or None
        self.metadata = dict(data.metadata)
        
        if data.include_content:
            self.prompt_content = data.prompt
            self.response_content = data.response
        else:
            self.prompt_content = None
            self.response_content = None
        
        self.receipt_hash = self._compute_receipt_hash()
        self._signature = ""
    
    @staticmethod
    def _hash_content(content: Any) -> str:
        """Hash any content using SHA-256 with JSON canonicalization"""
        if content is None:
            return canonicalize_and_hash(None)
        return canonicalize_and_hash(content)
    
    def _compute_receipt_hash(self) -> str:
        """Compute SHA-256 hash of the full receipt payload"""
        payload = {
            "version": self.version,
            "timestamp": self.timestamp,
            "session_id": self.session_id,
            "agent_id": self.agent_id,
            "prompt_hash": self.prompt_hash,
            "response_hash": self.response_hash,
            "scores": self.scores,
            "prev_receipt_hash": self.prev_receipt_hash,
            "metadata": self.metadata,
        }
        return canonicalize_and_hash(payload)
    
    async def sign(self, private_key: Union[str, bytes]) -> None:
        """Sign the receipt with an Ed25519 private key"""
        self._signature = sign(self.receipt_hash, private_key)
    
    @property
    def signature(self) -> str:
        """Get the signature (hex string)"""
        return self._signature
    
    @property
    def is_signed(self) -> bool:
        """Check if receipt is signed"""
        return len(self._signature) > 0
    
    async def verify(self, public_key: Union[str, bytes]) -> bool:
        """Verify the receipt's signature"""
        if not self._signature:
            return False
        return verify(self._signature, self.receipt_hash, public_key)
    
    def verify_chain(self, previous: Union['TrustReceipt', SignedReceipt]) -> bool:
        """Check if this receipt chains correctly from a previous receipt"""
        if isinstance(previous, TrustReceipt):
            prev_hash = previous.receipt_hash
        else:
            prev_hash = previous.receipt_hash
        return self.prev_receipt_hash == prev_hash
    
    def to_json(self) -> SignedReceipt:
        """Export as SignedReceipt object"""
        return SignedReceipt(
            version=self.version,
            timestamp=self.timestamp,
            session_id=self.session_id,
            agent_id=self.agent_id,
            prompt_hash=self.prompt_hash,
            response_hash=self.response_hash,
            scores=self.scores,
            prev_receipt_hash=self.prev_receipt_hash,
            receipt_hash=self.receipt_hash,
            signature=self._signature,
            metadata=self.metadata,
            prompt_content=self.prompt_content,
            response_content=self.response_content,
        )
    
    @staticmethod
    def from_json(data: SignedReceipt) -> 'TrustReceipt':
        """Create a TrustReceipt from JSON data"""
        dummy = TrustReceiptData(
            session_id=data.session_id,
            prompt="",
            response="",
            scores=data.scores,
        )
        receipt = TrustReceipt(dummy)
        
        receipt.timestamp = data.timestamp
        receipt.agent_id = data.agent_id
        receipt.prompt_hash = data.prompt_hash
        receipt.response_hash = data.response_hash
        receipt.prev_receipt_hash = data.prev_receipt_hash
        receipt.metadata = data.metadata
        receipt.receipt_hash = data.receipt_hash
        receipt._signature = data.signature
        receipt.prompt_content = data.prompt_content
        receipt.response_content = data.response_content
        
        return receipt
